packages/@lwc/wire-service/src/wiring.ts

```diff
@@ -168,18 +168,36 @@ export class WireEventTarget {
 
                 // setup listeners for all reactive parameters
                 const configContext = this._context[CONTEXT_ID][CONTEXT_UPDATED];
+                const reactiveParametersGroupByHead = new Map<string, Set<ReactiveParameter>>();
                 reactiveKeys.forEach(key => {
                     const reactiveParameter = buildReactiveParameter(reactives[key]);
-                    let configListenerMetadatas = configContext.listeners[reactiveParameter.head];
+                    const reactiveParameterHead = reactiveParameter.head;
+                    let configListenerMetadatas = configContext.listeners[reactiveParameterHead];
+
+                    if (!reactiveParametersGroupByHead.has(reactiveParameterHead)) {
+                        reactiveParametersGroupByHead.set(
+                            reactiveParameterHead,
+                            new Set<ReactiveParameter>()
+                        );
+                    }
+                    // reactiveParameter.head is set in the line above.
+                    const reactiveParametersWithSameHead = reactiveParametersGroupByHead.get(
+                        reactiveParameterHead
+                    )!;
```

Comment:
This code can be improved performance-wise and type safety-wise by avoiding the double map lookup:

```suggestion
                    let reactiveParametersWithSameHead = reactiveParametersGroupByHead.get(reactiveParameterHead);

                    if (isUndefined(reactiveParametersWithSameHead)) {
                        reactiveParametersWithSameHead = new Set<ReactiveParameter>();
                        reactiveParametersGroupByHead.set(reactiveParameterHead, reactiveParametersWithSameHead);
                    }
```